### PR TITLE
OFS-280: Contract - add signal for dynamic mutation like in PH

### PR DESCRIPTION
**Pull request by sniedzielski**
TICKET: https://openimis.atlassian.net/browse/OFS-280

In that PR I want to add processing signal after passing additional filter into payment tab - "contract".

It works even with both args passing into filters.
Screenshot from 2021-04-21 13-21-36
When I for example pass the contract id and policy holder id and policy holder is not related to the contract - the result of that query will be empty list.

### CODE REVIEW

File: contract/signals.py [hunk 1]
```diff
@@ -120,6 +125,30 @@ def activate_contracted_policies(sender, instance, **kwargs):
...
+            # check if we have interactive user from current context
+            if '(i)' in type_user:
+                ph_user = PolicyHolderUser.objects.filter(Q(policy_holder__id=contract_to_process.policy_holder.id, user__id=user.i_user.id)).first()
+                if ph_user:
+                    return Q(
+                        payment_details__premium__contract_contribution_plan_details__contract_details__contract__id=contract_id
+                    )
+
+
+signal_before_payment_query.connect(append_contract_filter)
```
Comment (delcroip) [inline contract/signals.py:149]: Small comment, we might want to group signal registration on the top of the page

File: contract/signals.py [hunk 2]
```diff
@@ -120,6 +125,30 @@ def activate_contracted_policies(sender, instance, **kwargs):
                         __save_or_update_contract(contract, contract.user_updated)
 
 
+# additional filters for payment in 'contract' tab
+def append_contract_filter(sender, **kwargs):
+    user = kwargs.get("user", None)
+    additional_filter = kwargs.get('additional_filter', None)
+    if "contract" in additional_filter:
```
Comment (delcroip) [inline contract/signals.py:132]: just check with @rstencel  that he will send "contract" and not "Contract"

File: contract/signals.py [hunk 3]
```diff
@@ -120,6 +125,30 @@ def activate_contracted_policies(sender, instance, **kwargs):
...
+            contract_id = additional_filter["contract"]
+            contract_to_process = Contract.objects.filter(id=contract_id).first()
+            # check if user is linked to ph in policy holder user table
+            type_user = f"{user}"
+            # related to user object output (i) or (t)
+            # check if we have interactive user from current context
+            if '(i)' in type_user:
+                ph_user = PolicyHolderUser.objects.filter(Q(policy_holder__id=contract_to_process.policy_holder.id, user__id=user.i_user.id)).first()
+                if ph_user:
+                    return Q(
```
Comment (delcroip) [inline contract/signals.py:144]: if ph_user or user.has_perms(PaymentConfig.gql_query_payments_perms)

--> so you filter also when there is the payment rights

**Review by delcroip — CHANGES_REQUESTED** [review]

File: contract/signals.py (hunk 1 above)
Comment (sniedzielski) [reply contract/signals.py:149]: ok, I will move above

File: contract/signals.py (hunk 2 above)
Comment (sniedzielski) [reply contract/signals.py:132]: ok. I will ask him about it

File: contract/signals.py (hunk 3 above)
Comment (sniedzielski) [reply contract/signals.py:144]: ok.

File: contract/signals.py (hunk 2 above)
Comment (sniedzielski) [reply contract/signals.py]: @delcroip Ralal has  just confirmed me that "contract". Therefore I don't have to change this.

**Comment by sniedzielski** [comment]
@delcroip You can re-check this. I made some changes

**Review by delcroip — APPROVED** [review]

**Pull request merged**